### PR TITLE
fix: stuck keys on leave

### PR DESCRIPTION
**Pull request by enzious**
Blocked by: https://github.com/deskflow/deskflow/pull/7522

---

On exiting a screen, deskflow attempts to unset keys that are down while leaving the screen. In some cases, such as with Wayland (EiScreen), the keys are reset but after the device emulation has already ended by `Screen::leave()`. This leaves the keys in the pressed state since the reset will fail after emulation stops, and may leave the keys stuck pressed until the environment implementing libei is restarted (such as with Kwin).

This pull request splits `Screen::leave()` into `Screen::canLeave()` and `Screen::leave()`, allowing for the keys to be reset in between.

Basically:
Check for screen leave allowed -> Reset keys -> Actually leave

Resolves: #7518

### CODE REVIEW

**Comment by enzious** [comment]
Side note - after rebasing on master, I'm getting the following error while attempting to build for Windows:

image

File: src/lib/platform/XWindowsScreen.cpp [hunk 1]
```diff
@@ -326,7 +326,16 @@ void XWindowsScreen::enter() {
   m_isOnScreen = true;
 }
 
-bool XWindowsScreen::leave() {
+bool XWindowsScreen::canLeave() {
+  // see if grabbing the mouse and keyboard, if primary, is possible
+  if (m_isPrimary && !grabMouseAndKeyboard()) {
```
Comment (sithlord48) [inline src/lib/platform/XWindowsScreen.cpp:331]: This should be this same logic.
`return !(m_isPrimary && !grabMouseAndKeybaord());`

**Review by sithlord48 — COMMENTED** [review]
The initial commit can be dropped as its just changed completely by the second one. 
Otherwise looks pretty good. Testing on inputleap

**Comment by nbolton** [comment]
Thanks for the PR, we're currently renaming the project and I'm in the process of restoring CI functionality.
Updates: https://github.com/deskflow/deskflow/discussions/7517
Once CI is working, we can get your PR merged in.

File: src/lib/platform/XWindowsScreen.cpp [hunk 2]
```diff
@@ -326,7 +326,12 @@ void XWindowsScreen::enter() {
   m_isOnScreen = true;
 }
 
-bool XWindowsScreen::leave() {
+bool XWindowsScreen::canLeave() {
```
[the remark this answers is not in the document]
Comment (enzious) [reply src/lib/platform/XWindowsScreen.cpp:329]: Interesting, I'll have to test this after work.

[the remark this answers is not in the document]
Comment (enzious) [reply src/lib/platform/XWindowsScreen.cpp:329]: Actually, fix was pretty simple. Pushed it.

File: src/lib/platform/EiScreen.cpp [hunk 3]
```diff
@@ -345,7 +345,9 @@ void EiScreen::enter() {
 #endif
 }
 
-bool EiScreen::leave() {
+bool EiScreen::canLeave() { return true; }
```
Comment (sithlord48) [inline src/lib/platform/EiScreen.cpp:348]: Future improvement: This should be our default from the base class

Comment (nbolton) [reply src/lib/platform/EiScreen.cpp:348]: Perhaps it's worth making an issue for this.

Comment (enzious) [reply src/lib/platform/EiScreen.cpp:348]: Should I just resolve it here in this PR?

**Review by sithlord48 — APPROVED** [review]
